etc/make_small_lib.pl

```diff
@@ -0,0 +1,335 @@
+    close $src;
+
+    $filename = basename($filename);
+    $filename =~ s/^bn_|\.c$//g;
+
+    $content =~ s{/\*.*?\*/}{}gs;
+    my $list = "";
+    foreach my $line (split /\n/, $content) {
+      # TODO: change to read public functions only when scanning user source.
+      while ($line =~ /(fast_)?(s_)?mp\_[a-z_0-9]*(?=\()|(?<=\()mp\_[a-z_0-9]*(?=,)/g) {
+        my $a = $&;
+        next if $a eq "mp_err";
+        if($list eq "") {
+          $list = $a;
+        }
+        else {
+          $list = $list . "," . $a;
+        }
+      }
+    }
+    $depmap{$filename} = $list;
+  }
+  return %depmap;
+}
+
+sub gather_dependencies
+{
+   my ($deplist, $depmap, $funcslist) = @_;
+   my @funcs = split ',', $funcslist;
+   if ($deplist =~ /$funcs[0]/) {
+      return $deplist;
+   } else {
+      $deplist = $deplist . $funcs[0];
+   }
+   push @dependency_list, $funcs[0];
+   shift @funcs;
+   my $olddeplist = $deplist;
+   foreach my $i (@funcs) {
+      $deplist = gather_dependencies($deplist, $depmap, ${$depmap}{$i}) if exists ${$depmap}{$i};
+   }
+   return $olddeplist;
+}
+
+sub start
+{
+  my ($sd, $td, $no, $cm) = @_;
+
+  my %depmap;
+  my %user_functions;
+  my %ff_hash;
+  my @functions;
+  my @tmp;
+
+  # TODO: checks&balances
+   -e $td.$sep."tommath.h" or die "$td.$sep.tommath.h not found, please check path to LibTomMath sources\n";
+
+  %depmap = gather_functions($td);
+  %user_functions = gather_functions($sd);
+
+  foreach (sort keys %user_functions) {
+    push @functions, split /,/ , $user_functions{$_};
+  }
+  @functions = uniq(sort @functions);
+
+  foreach (@functions) {
+    exists $depmap{$_} or die "Function \"$_\" does not exist in LibTomMath.\n";
+  }
+
+  # gather dependencies
+  foreach (sort keys %user_functions) {
+    gather_dependencies("", \%depmap, $user_functions{$_});
+  }
+  @dependency_list = uniq(sort @dependency_list);
+
+  # make an even smaller lib by removing non-essential functions (e.g. Karatsuba for multiplication)
+  if ($no == 1) {
+    foreach (@dependency_list) {
+      next if exists $fast_functions{$_};
+      push @tmp, $_;
+    }
+    @dependency_list = @tmp;
+    @tmp = ();
+  }
+
+  if ($cm == 1) {
+    foreach my $entry (@dependency_list) {
+      $entry = 'bn_' . $entry . '.o';
+      push @tmp, $entry;
+    }
+    process_makefiles($td, @tmp);
+  }
+  else {
+    write_header($td, @dependency_list);
+  }
+  return 1;
+}
+
+sub die_usage {
+  die <<"EOO";
+usage: $0 -s   OR   $0 --source-dir         [./]
+       $0 -t   OR   $0 --tommath-dir        [./libtommath]
+       $0 -n   OR   $0 --no-optimization
+       $0 -m   OR   $0 --change-makefiles
+
+The option --source-dir accepts a directory or a single file.
+
+EOO
+}
+
+my $source_dir = "";
+my $tommath_dir = "libtommath$sep";
+my $config_file = "";
+my $no_optimizations = 0;
+my $change_makefiles = 0;
+
+GetOptions( "s|source-dir=s"        => \$source_dir,
+            "t|tommath-dir=s"       => \$tommath_dir,
```

Comment:
Could we avoid the over configuration? Just always run in ./

Reply:
> Just always run in ./

That's the default and it was cheap to offer options.

You can also install it and use if for different versions of LTM. Not useful now but once a new version comes out where all the deprecated stuff is actually gone you might need to keep older versions for legacy reasons.
And don't ignore the legacy part, there are people out there who make a nice penny from supporting old software and even bought an old laptop from ebay to get a fully functional parport to be able to read the backups stored with zipdrives ;-)

The goal of this script has changed a little bit. It is now a freestanding tool to adapt LTM to the user's needs only, like @sjaeckel suggested. It is also a bit more "Unix-y" this way.

Reply:
> The goal of this script has changed a little bit. It is now a freestanding tool to adapt LTM to the user's needs only, like @sjaeckel suggested. It is also a bit more "Unix-y" this way.

Why do we need this as a freestanding tool? I don't get that. Did @sjaeckel order this? ;)

Reply:
> Why do we need this as a freestanding tool?

To part the tools for the LTM developers from the tools for the users of LTM. The latter is gathered in `etc` for now and the first one in `helper.pl`.
It is not the best way to do it but the organisation in LTM is a bit odd. The documentation is in `doc`, the logs are in `logs`, in `mtest` is the stuff necessary for that test and a copy of Michael Fromberger's `mpi.c`, in `demo` is the testing stuff but no demo, and `etc` is quite a mixed bag. All in all a real mess. The organization scheme common in many larger source-trees `src`, `build`, `docs`, `scripts`, and `tests` with the configuration stuff in the root directory is not that bad and can be used for LTM, too.

But that's a *lot* of work with no gain but that satisfied feeling once it's done and until than&hellip;

> Did @sjaeckel order this?

I'm no-one's servant nor any-one's master; will neither give nor take orders.

Unless you pay me well, of course, such that I can pay a professional to look over my english grammar and point out all the ambiguous parts in my texts, too.

Reply:
;)

Reply:
> Why do we need this as a freestanding tool?

I'm seeing this as the prototype-phase w/o interfering with helper.pl too much (less merge conflicts and stuff) so it's fine like that

> Did @sjaeckel order this? ;)

lol

no but I would've started like that as well I guess ;)

Reply:
@czurnieden Just to clarify, this was a joke. Sorry if you got offended. Or maybe I got caught by you joking back...

@sjaeckel I agree fully to have it freestanding for the prototyping phase. But above, @czurnieden said 
> The goal of this script has changed a little bit. It is now a freestanding tool to adapt LTM to the user's needs only

Reply:
> Just to clarify, this was a joke.

You put a smiley at the end, so&hellip;?

> Sorry if you got offended

That's nigh impossible, that bar hangs *very* high, it needs a lot to make this friendly bear remembering that a bear is omnivore.
And also: as somebody who dishes out with both hands I can hardly complain if I have to take some, too ;-)

> Or probably I got caught by you joking back...

Yeah, I'm too complicated sometimes. The play with the increasingly limping metre in "I'm no-one's servant&hellip;" was probably a bit over the top?

But back to biz:

> @sjaeckel I agree fully to have it freestanding for the prototyping phase. But above, @czurnieden said "The goal of this script has changed a little bit. It is now a freestanding tool to adapt LTM to the user's needs only"

Yes, we have different points of view. I don't see it as a bad thing. especially if the outcome is the same ;-)

I named the functions I copied from `helper.pl` or `gen.pl` differently if I changed them and kept their names otherwise, as you would expect. With the exception of some clashes in the short options it should be no problem to put them all in `helper.pl` in about half an hour, including the essential coffee break. It would be a much bigger effort the other way around.

Or to put it more bluntly: insisting on putting it all into `helper.pl` *now* comes too early.